### PR TITLE
ImageAnimatorWCS Update 

### PR DESCRIPTION
**Pull request by prateekiiest**
User can now define the units for ImageAnimatorWCS. Default will be u.km.

- [ ] User defined inputs 
- [ ] Ticks for ImageAnimator

Should it be something like taking the user input for the unit from keyboard or let the user define the unit while calling the function ?

#2288

### CODE REVIEW

**Comment by nabobalis** [comment]
There are two issues so far: 

1. We should allow user input for units but not have default units since we can not be 100% sure those defaults will apply. 
2. The arguments order in the function call are in the wrong order, it is while all the CI has failed.

**Comment by prateekiiest** [comment]
so some sort of `z = raw_input()` from the user side.
And then `Unit(z)` to convert to specified astropy units ?

**Comment by nabobalis** [comment]
Well personally, I am not a fan of using user input. It should be changed using the function call, like you set it now.

Enforcing the units to be astropy units I have no problem with but it's that setting the default  to be distance will not be 100% right.

Also, I think it would be safe to assume that people will not give out a unit 100% of the time and still handle the lack of a unit being provided.

**Comment by nabobalis** [comment]
The order of the inputs are still out of place.

File: sunpy/visualization/imageanimator.py [hunk 1]
```diff
@@ -936,7 +937,9 @@ class ImageAnimatorWCS(ImageAnimator):
     Extra keywords are passed to imshow.
 
     """
-    def __init__(self, data, wcs=None, image_axes=[-1, -2], unit_x_axis=None, unit_y_axis=None,
+    
+    
+    def __init__(self, data, wcs=None, image_axes=[-1, -2], unit_x_axis, unit_y_axis,
```
Comment (nabobalis) [inline sunpy/visualization/imageanimator.py]: Here the order is broken.

File: sunpy/visualization/imageanimator.py [hunk 2]
```diff
@@ -949,20 +952,20 @@ def __init__(self, data, wcs=None, image_axes=[-1, -2], unit_x_axis=None, unit_y
         self.slices_wcsaxes = list_slices_wcsaxes[::-1]
         self.unit_x_axis = unit_x_axis
         self.unit_y_axis = unit_y_axis
-        super(ImageAnimatorWCS, self).__init__(data, image_axes=image_axes, axis_ranges=axis_ranges, **kwargs)
+        super(ImageAnimatorWCS, self).__init__(data, image_axes=image_axes, unit_x_axis, unit_y_axis , axis_ranges=axis_ranges, **kwargs)
```
Comment (nabobalis) [inline sunpy/visualization/imageanimator.py]: Here the order is broken.

Comment (prateekiiest) [reply sunpy/visualization/imageanimator.py]: But the relative ordering is same as before it was updated?

Comment (nabobalis) [reply sunpy/visualization/imageanimator.py]: In Python, it goes required inputs, inputs with default/given values like  ``image_axes=image_axes``

Comment (prateekiiest) [reply sunpy/visualization/imageanimator.py]: ok I got it

**Comment by prateekiiest** [comment]
@nabobalis 
how about something like this for the user input:

`if(unit_x_axis is not None):
        unit_x_axis =u.Unit(raw_input())
  if(unit_y_axis is not None):
       unit_y_axis = u.Unit(raw_input())`

**Comment by nabobalis** [comment]
@prateekiiest Personally, I do not like using raw_input/input.  It feels out of place in how someone will use this in their scripts. 

I think the angle we want to take would be to create a new function that allows a user to call it and set the units/labels that way instead of using the ``__init__``. A big reason might be that we could have N axes that someone will want to label, but right now I am not 100% sure on this.

**Comment by nabobalis** [comment]
So can you please explain to me how I would use this new function in an example?

**Comment by prateekiiest** [comment]
Yeah sure,
For example if we see this function

`def _get_unit_from_user(unit_x_axis = None, unit_y_axis = None):`
        
        if unit_x_axis is not None:
            unit_x_axis = u.Unit(unit_x_axis)
        if unit_y_axis is not None:
            unit_y_axis = u.Unit(unit_y_axis)


As you said user may supply the input or may not.
If the user supplies the input i.e. unit_x_axis = 'meter' and unit_y_axis = 'second' for example
`unit_x_axis = u.Unit(unit_x_axis)` will be converted to the corresponding astropy units.

`u.Unit('meter') = u.m`
`u.Unit('second') = u.s`

So we can let the user give any input (not necessarily u,m or u.s) - just simple meter/second / kg/km we can convert it to the corresponding astropy units without any overhead for the user.

**Comment by prateekiiest** [comment]
@nabobalis  any suggestions on this you would like to give ?

**Comment by nabobalis** [comment]
Well, the function name I have two issues with. Underscore names are typically private functions called within code and not by a user. Secondly, the name itself is kind of a bit verbose. 

If this function need to be called by a user, I do not think it needs to have any defaults. To my mind, it should error if the user does not input any units.

**Comment by prateekiiest** [comment]
so how should we handle defaults then and also enabling user -defined inputs?

**Comment by nabobalis** [comment]
@prateekiiest Did you see the comments from @DanRyanIrish [here?](https://github.com/sunpy/sunpy/issues/2288#issuecomment-339453136)

I think as it currently stands, this does not address the original open issue.

**Comment by prateekiiest** [comment]
Yeah @nabobalis , I was working on some part of it a few days back. Its not complete yet and I am thinking about how to ensure the ticks are appropriate.

**Comment by DanRyanIrish** [comment]
Hi @prateekiiest and @nabobalis.  Just catching up on this thread.  My initial thought for a solution was that the ```ImageAnimatorWCS``` would include optional ```unit_x_axis=None``` and ```unit_y_axis=None``` kwargs (or other appropriate names) in its ```__init__```.  If they were not set, they would default to the units in the WCS object itself.  This would ensure that the units are always of appropriate type.  If there's a good reason to have a separate public function called by the user to define the units, I'm open to it.  But the API I suggested above introduces the least complexity from the users' point of view.  The necessary unit manipulation could be done in a private function if needs be, called by ```ImageAnimatorWCS```.

I'm OK with forcing the input to be astropy units, although it would be nicer to allow valid unit strings as well as you guys discussed.

**Comment by prateekiiest** [comment]
Thanks @DanRyanIrish  for the clear view

File: sunpy/visualization/imageanimator.py [hunk 3]
```diff
@@ -954,11 +957,25 @@ def __init__(self, data, wcs=None, image_axes=[-1, -2], unit_x_axis=None, unit_y
         list_slices_wcsaxes[image_axes[0]] = 'x'
         list_slices_wcsaxes[image_axes[1]] = 'y'
         self.slices_wcsaxes = list_slices_wcsaxes[::-1]
+
+        if(unit_x_axis is not None):
+            self.unit_x_axis = u.Unit(unit_x_axis)
```
Comment (prateekiiest) [inline sunpy/visualization/imageanimator.py]: If the user input is not None, then its get converted to the astropy units.
Like for any naive input for example `meters` , `self.unit_x_axis = u.m` .

File: sunpy/visualization/imageanimator.py [hunk 4]
```diff
@@ -969,6 +986,7 @@ def _set_unit_in_axis(self, axes):
         if self.unit_x_axis is not None:
             axes.coords[2].set_format_unit(self.unit_x_axis)
```
Comment (prateekiiest) [inline sunpy/visualization/imageanimator.py]: so the axes units will be set to the unit of the `unit_x_axis` which is `u.m`.

**Comment by prateekiiest** [comment]
I consulted this documentation https://media.readthedocs.org/pdf/wcsaxes/latest/wcsaxes.pdf for the axes features.

@DanRyanIrish @nabobalis , let me know if this is the thing that's desired

**Comment by nabobalis** [comment]
Good question. It's been a while so my memory is kind of fuzzy. We got to the point where we can add units and they get labelled now?

**Comment by DanRyanIrish** [comment]
Hi @prateekiiest.  Thanks for the updates to this PR.  I missed them in December as I was travelling and then it was holidays.  I will take a look some time this week.

**Comment by prateekiiest** [comment]
For the current implementation , I worked with setting the units of the corresponding user units.
Followed the above pdf file for reference

**Comment by nabobalis** [comment]
@prateekiiest Can you create a short example to show that it works?

File: sunpy/visualization/imageanimator.py [hunk 5]
```diff
@@ -954,11 +957,25 @@ def __init__(self, data, wcs=None, image_axes=[-1, -2], unit_x_axis=None, unit_y
...
+            self.unit_y_axis = u.Unit(unit_y_axis)
+        else:
+            self.unit_y_axis = unit_y_axis            
+       
+        super(ImageAnimatorWCS, self).__init__(data, image_axes=image_axes, axis_ranges=axis_ranges, **kwargs)
+
         self.unit_x_axis = unit_x_axis
         self.unit_y_axis = unit_y_axis
         super(ImageAnimatorWCS, self).__init__(data, image_axes=image_axes,
                                                axis_ranges=axis_ranges, **kwargs)
```
Comment (DanRyanIrish) [inline sunpy/visualization/imageanimator.py]: I assume the above three lines should be deleted in light of the code block above?

**Comment by wafels** [comment]
See http://docs.astropy.org/en/stable/visualization/wcsaxes/ticks_labels_grid.html#tick-label-format

**Comment by Cadair** [comment]
I really don't like the API of the units as keywords to the construction of `WCSImageAnimator`.

I propose this API:

```python
wani = WCSImageAnimator(...)
wani.set_image_axes_units(u.arcsec, u.arcsec)
```

Where `set_image_axes_units` takes a two positional arguments and then it uses the `image_axes` keyword argument to set the **correct** coordinate axes on the plot.

**Comment by prateekiiest** [comment]
@nabobalis , I guess we should first work on the API as suggested by @Cadair.

If it works, I can then start with creating an example based on that API.
I didn't get time yesterday, I will sit with the code today.

**Comment by prateekiiest** [comment]
shifted the units to a new function `set_image_axes_units`

**Comment by prateekiiest** [comment]
@Cadair something like this ?

**Comment by nabobalis** [comment]
Generally, a function that starts with an underscore is not used outside of the code itself, it shouldn't be called by a user.

**Comment by prateekiiest** [comment]
OK, I get it it should be called by the user itself instead of that being done by the imageanimatorWcs itself.

**Comment by nabobalis** [comment]
Could we see a code example and the output?

**Comment by prateekiiest** [comment]
any example for data value and wcs value ?
something like `data=map.data`, `wcs=map.wcs`

**Comment by nabobalis** [comment]
Oh no, I meant an example that demonstrates the new additions with its output kind of like how #2442 does it.

**Comment by prateekiiest** [comment]
i know that @nabobalis , but dont we need to pass the arguments for `data` and `wcs` in the `imageanimator` function.

**Comment by nabobalis** [comment]
Sure? I'm not sure I follow.

**Comment by prateekiiest** [comment]
@nabobalis I worked with multiple fits files  for the MapcubeAnimator.

Since the flow is like this

`ImageAnimatorWCS <- ImageAnimator Instance <- ArrayAnimator <- BaseFuncAnimator <- MapCubeAnimator Instance ` .

Once MapcubeAnimator.updatefig doesn't do the right thing , I am not able to produce BaseFunc and the rest in the hierarchy.

#2444 Same problem I get for Multiple fits file

**Comment by nabobalis** [comment]
Sorry, I'm even more confused.

**Comment by prateekiiest** [comment]
How do I get to see the slider functions of a mapcube.peek ? 
`ani = mapcube.peek('...')`

`ani.slider_funtions? `

**Comment by nabobalis** [comment]
I think that slider_functions is just a call to update_fig. I am not sure it was ever made to be accessed outside of the class.

**Comment by prateekiiest** [comment]
I guess `BaseFuncAnimator.plot_start_image` is not implemented

**Comment by prateekiiest** [comment]
And I dont think `update_fig `uses `slider` parameter anywhere in the code that I went through here
http://docs.sunpy.org/en/v0.8.3/_modules/sunpy/visualization/mapcubeanimator.html#MapCubeAnimator.updatefig

**Comment by nabobalis** [comment]
No it doesn't look like it. I wonder if it gets used internally somehow.

**Comment by prateekiiest** [comment]
Then why do we need to pass that `slider` value separately? Its no use, I assume

**Comment by nabobalis** [comment]
Not sure, could remove it and see if the function still works. 

But I am unsure what this all has to do with the current PR.

**Comment by prateekiiest** [comment]
I said earlier too.

The visualization code is like a hierarchy.

Since ImageAnimatorWCS uses instance of ImageAnimator which in turn uses instance of Arrayanimator and which finally uses the BaseFuncAnimator.

So until I can't create a BaseFunc example, how do I pass the instances all the way down to ImageWCS?

**Comment by nabobalis** [comment]
I don't understand what you mean by passings the instances all the way down to ImageWCS. 
Or what this has to do with setting the units/labels of the plot?

**Comment by prateekiiest** [comment]
I guess all these classes are derived classes. Like ImageAnimatorWCS is  derived from ImageAnimator.
So I was might have confused you about the instance being used in this classes.

I am getting the following error when I am running the code.

        from matplotlib import pyplot as plt
        from astropy.io import fits
        from astropy.wcs import WCS
        from astropy.utils.data import get_pkg_data_filename

        #filename = get_pkg_data_filename('tutorials/FITS-images/HorseHead.fits')

        from astropy.utils.data import get_pkg_data_fileobj
        with get_pkg_data_fileobj('allsky/allsky_rosat.fits',
                                   encoding='binary') as fobj:  
             fcontents = fobj.read()
        hdu = fits.open(fobj)[0]
        wcs = WCS(hdu.header)
        ImageAnimatorWCS(data = hdu.data, wcs=wcs)

AttributeError: 'ImageAnimatorWCS' object has no attribute 'slices_wcsaxes'

Is there anything that I am doing wrong here?

**Comment by nabobalis** [comment]
I just ran that code and got:

```
 ax = self.sliders[ind]
IndexError: list index out of range```

**Comment by prateekiiest** [comment]
I believe `self.sliders` comes out to be an empty list.
What is the value of `ind` that you passed?

**Comment by nabobalis** [comment]
I only ran that code. I assume its an issue with the input data being 2D again.

**Comment by prateekiiest** [comment]
could you please give me an example for the wcs data where naxis =1?

**Comment by nabobalis** [comment]
I can't seem to figure out the right WCS to pipe into the function.
@Cadair some help please.

**Comment by prateekiiest** [comment]
@nabobalis , I guess the docs need to be updated in case  if  ImageanimatorWCS only accepts 1D data

http://docs.sunpy.org/en/stable/api/sunpy.visualization.imageanimator.ImageAnimatorWCS.html#sunpy.visualization.imageanimator.ImageAnimatorWCS

**Comment by nabobalis** [comment]
I'm sorry I don't follow. It animates Nd data.

**Comment by prateekiiest** [comment]
oh! my bad. 

But the data that will be passed into the function , will that data not be animated ?

**Comment by nabobalis** [comment]
It should be. It is what Mapcubeanimtor uses

**Comment by prateekiiest** [comment]
do we need to explicitly pass the `button_func` and `button_list` in the above function?

**Comment by nabobalis** [comment]
You mean into ImageAnimatorWCS?

**Comment by prateekiiest** [comment]
yes

**Comment by nabobalis** [comment]
I'm aware it needs to be. Those should be created automatically if I recall.

**Comment by prateekiiest** [comment]
So what's your opinion, are there any current bug that needs to be fixed first before going forward ?

**Comment by nabobalis** [comment]
You have to start from the beginning, I've been lost for a while now.

What is the bug and what's the aim going forward?

**Comment by prateekiiest** [comment]
`Problem` : data and wcs to be passed must be 1D , otherwise function throws error.

`Aim` : Make plot ticks and other features for the function

**Comment by nabobalis** [comment]
I don't get why we are talking about 1D data. I believe we have a LineAnimator for that kind of data. 

ImageAnimator does assume 2D or greater, as `image_axes=[-1, -2]` is set as two axes.

**Comment by prateekiiest** [comment]
but it shows the same error as you said in the previous comments using 2D data.

The example code I posted above

**Comment by nabobalis** [comment]
Basically in that case you only have one 2D data. There is no third axis to animate over. 

Mapcube animator is a wrap around this this code and it works for a mapcube that has more than one file.

**Comment by nabobalis** [comment]
So I got it to finally work once I copied a basic WCS dictionary from `ndcube`.

Files here are just some SDO files I have.

```python
cube = smap.Map(files[0:10],cube=True)
cube = cube.as_array()
# So the last two axes are the image
# The shape is (10, 4096, 4096)
cube = np.rollaxis(cube,-1)

wcs_input_dict = {
'CTYPE1': 'Time    ', 'CUNIT1': 's', 'CDELT1': 12,
'CTYPE2': 'HPLN-TAN', 'CUNIT2': 'arcsec', 'CDELT2': 0.6, 
'CTYPE3': 'HPLT-TAN', 'CUNIT3': 'arcsec', 'CDELT3': 0.6}

WCS = astropy.wcs.WCS(wcs_input_dict)

ImageAnimatorWCS(cube, wcs=WCS, vmax=1000)
```

Gives me a plot that moves in time via the slider.


Changing it to,
```python
cube = smap.Map(files[0],cube=True)
cube = cube.as_array()
# The shape is (1, 4096, 4096)
cube = np.rollaxis(cube,-1)
```
Gives me an image that like before but not does not change.

If I remove that empty axis, so the data is now `(4096, 4096)`,  I get
```python
ax = self.sliders[ind]
IndexError: list index out of range
```

This is should be expected.

**Comment by prateekiiest** [comment]
I am still getting overflow error as per your code.

```
 imin = np.ceil(value_min / spacing)
     98         imax = np.floor(value_max / spacing)
---> 99         values = np.arange(imin, imax + 1, dtype=int)
    100         return values
    101 

OverflowError: Python int too large to convert to C long

```
Its exceeding `sys.maxsize`

**Comment by nabobalis** [comment]
You are creating arrays or numbers too large for your system. I assume that is a 32bit limit.

**Comment by prateekiiest** [comment]
how do I get about this. I never actually came across such issue before

**Comment by nabobalis** [comment]
What is the code you are trying to run?

**Comment by prateekiiest** [comment]
yours only. I had creating around 8 maps using sunpy.map and then passing into the mabcube

**Comment by prateekiiest** [comment]
breaks at the ImageAnimatorWCS and um, I am using the AIA_193 IMAGES

**Comment by nabobalis** [comment]
I just used that and had no problems.

What is the output of sunpy.util.system_info()?

**Comment by prateekiiest** [comment]
```
==========================================================
 SunPy Installation Information

==========================================================

###########
 General
###########
Time : Tuesday, 13. February 2018 01:27PM UT
System : Windows
Processor : Intel64 Family 6 Model 61 Stepping 4, GenuineIntel
Arch : 64bit
SunPy : 0.8.3
SunPy_git : 5437856a2ab0bff996b6a8740f38d3631cca642d
OS: Windows 8.1 6.3.9600 (Intel64 Family 6 Model 61 Stepping 4, GenuineIntel)


###########
 Required Libraries 
###########
Python: 2.7.12
NumPy: 1.11.1
SciPy: 0.17.1
matplotlib: 1.5.1
Astropy: 2.0.2
Pandas: 0.18.1


###########
 Recommended Libraries 
###########
beautifulsoup: 4.4.1
PyQt: 4.11.4
SUDS: 0.6
Sqlalchemy: 1.0.13
Requests: 2.13.0
```

**Comment by nabobalis** [comment]
So you have a 64bit system, you are using anaconda?

What is your `sys.maxsize`?

**Comment by prateekiiest** [comment]
9223372036854775807L

**Comment by nabobalis** [comment]
I honestly have no idea. I thought that somehow it was a 32bit issue. But your system is fine.

**Comment by prateekiiest** [comment]
That's the thing. Lemme see if I can do anything with it.

**Comment by nabobalis** [comment]
Oh that might be an issue with python 2. I would try and create a python 3 environment. Also alot of the packages are quite of of date.

**Comment by nabobalis** [comment]
Python 2 seems to have a max int, which should be under `sys`, Python 3 removed that limit.

**Comment by prateekiiest** [comment]
ok, I get it. I suppose you did it in Py3

**Comment by nabobalis** [comment]
I do have a py2 one I can test with. Please hold.

**Comment by nabobalis** [comment]
It works under Py2 as well. Maybe just an update of the python environment would be enough.

**Comment by prateekiiest** [comment]
updated but still the same error

**Comment by nabobalis** [comment]
All I could suggest is try Py3 but I don't know if that will fix your problem.

**Comment by prateekiiest** [comment]
I tried all possible means to work out with this error, but I can't figure it why I am getting this error.
@nabobalis, will it possible for you if you can try with my code(in the PR) in your machine and see if the output works?

**Comment by nabobalis** [comment]
Sure, no problem here. 

Did you try Py3?

**Comment by prateekiiest** [comment]
Well, I tried to fix this from Py2 only, searched over stackoverflow and all.  I didn't shift to py3 since I got an error while importing the astropy.wcs (we discussed in riot) using py3.  So had to reinstall py version 2,7 and it worked then . Shifting to py3 may again give some unknown issue since I am not using Anaconda v3.

**Comment by nabobalis** [comment]
If you have time, I would probably start with miniconda from scratch.

**Comment by nabobalis** [comment]
What happened with this PR?

**Comment by prateekiiest** [comment]
I had to work with anaconda for the time being for the Mapcubes example because all the current programs are being used. I tried about finding the error but got no clue as to why this occurred. So, once the other PR gets finished, I will try with miniconda to see if the error goes away, but I am still not sure though.

**Comment by prateekiiest** [comment]
I currently installed anaconda on Ubuntu in my Virtual Machine.
I will see if the problem can be addressed there, hopefully I will sit down with with after 20th (semesters going on).

I will also try to update with necessary functionalities as needed by the issue. This current PR has not addresses the issue till yet and its been open for so many days without any productivity. I will try to finish this soon as possible, after this problem goes away.

Sorry for the so long delay @DanRyanIrish.

**Comment by nabobalis** [comment]
@prateekiiest You had no more time to work on this?

**Comment by prateekiiest** [comment]
After 9:30 today

**Comment by prateekiiest** [comment]
Need to resolve the conflicts first and rebase this

**Comment by prateekiiest** [comment]
I guess I have to a new PR for this.
There are too many conflicts to be resolved, first. And its not helping much using `mergetool`

**Comment by nabobalis** [comment]
You are using meld to as the git mergetool?

**Comment by nabobalis** [comment]
If it is too much to rebase, yeah I guess a new PR is the only way.

**Comment by prateekiiest** [comment]
no vimdiff

**Comment by nabobalis** [comment]
I wouldn't recommend vimdiff.

**Comment by prateekiiest** [comment]
I will be opening a new PR for this, and I will link to this before closing this down

**Pull request closed without merging**